--- src/de/impls.rs
+++ src/de/impls.rs
@@ -421,9 +421,10 @@
             let result =
                 super::impl_core::collect_into_array(&mut (0..N).map(|_| T::decode(decoder)));

-            // result is only None if N does not match the values of `(0..N)`, which it always should
-            // So this unwrap should never occur
-            result.unwrap()
+            match result {
+                | Some(r) => r,
+                | None => crate::error::cold_decode_error_array_length_mismatch(N, 0),
+            }
         }
     }
 }
@@ -501,9 +502,10 @@
                 &mut (0..N).map(|_| T::borrow_decode(decoder)),
             );

-            // result is only None if N does not match the values of `(0..N)`, which it always should
-            // So this unwrap should never occur
-            result.unwrap()
+            match result {
+                | Some(r) => r,
+                | None => crate::error::cold_decode_error_array_length_mismatch(N, 0),
+            }
         }
     }
 }
